File: app/src/main/java/com/arman/queuetube/util/itemtouchhelper/VideoItemTouchHelper.java
package com.arman.queuetube.util.itemtouchhelper;

import android.graphics.Canvas;

import com.arman.queuetube.util.itemtouchhelper.adapters.ItemTouchHelperAdapter;
import com.arman.queuetube.util.itemtouchhelper.viewholders.ItemTouchHelperViewHolder;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

public class VideoItemTouchHelper extends ItemTouchHelper {

    public VideoItemTouchHelper(@NonNull Callback callback) {
        super(callback);
    }

    public static class Callback extends ItemTouchHelper.Callback {

        private ItemTouchHelperAdapter adapter;

        private int dragFromIndex, dragToIndex;

        public Callback(ItemTouchHelperAdapter adapter) {
            this.adapter = adapter;
            this.dragFromIndex = -1;
            this.dragToIndex = -1;
        }

        @Override
        public boolean isLongPressDragEnabled() {
            return true;
        }

        @Override
        public boolean isItemViewSwipeEnabled() {
            return false;
        }

        @Override
        public boolean canDropOver(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder current, @NonNull RecyclerView.ViewHolder target) {
            return current.getItemViewType() == target.getItemViewType();
        }

        @Override
        public int getMovementFlags(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
            int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
            int swipeFlags = ItemTouchHelper.START | ItemTouchHelper.END;
            return makeMovementFlags(dragFlags, swipeFlags);
        }

        @Override
        public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
            if (viewHolder.getItemViewType() != target.getItemViewType()) {
                return false;
            }

            int fromIndex = viewHolder.getAdapterPosition();
            int toIndex = target.getAdapterPosition();

            if (this.dragFromIndex < 0) {
                this.dragFromIndex = fromIndex;
            }
            this.dragToIndex = toIndex;

            this.adapter.onItemDragged(viewHolder.getAdapterPosition(), target.getAdapterPosition(), false);
            return true;
        }

        @Override
        public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
            this.adapter.onItemSwiped(viewHolder.getAdapterPosition());
        }

        @Override
        public void onChildDraw(@NonNull Canvas c, @NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, float dX, float dY, int actionState, boolean isCurrentlyActive) {
            if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
                float alpha = 1f - Math.abs(dX) / (float) viewHolder.itemView.getWidth();
                viewHolder.itemView.setAlpha(alpha);
                viewHolder.itemView.setTranslationX(dX);
            } else {
                super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
            }
        }

        @Override
        public void onSelectedChanged(@Nullable RecyclerView.ViewHolder viewHolder, int actionState) {
            if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
                if (viewHolder instanceof ItemTouchHelperViewHolder) {
                    ((ItemTouchHelperViewHolder) viewHolder).onItemSelected();
                }
            }

            super.onSelectedChanged(viewHolder, actionState);
        }

        @Override
        public void clearView(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
            super.clearView(recyclerView, viewHolder);

            if (this.dragFromIndex >= 0 && this.dragFromIndex != this.dragToIndex) {
                this.adapter.onItemDragged(this.dragFromIndex, this.dragToIndex, true);
            }
            this.dragFromIndex = this.dragToIndex = -1;

            viewHolder.itemView.setAlpha(1f);
            if (viewHolder instanceof ItemTouchHelperViewHolder) {
                ((ItemTouchHelperViewHolder) viewHolder).onItemClear();
            }
        }

    }

}

File: app/src/main/java/com/arman/queuetube/util/itemtouchhelper/VideoItemTouchHelper.kt
package com.arman.queuetube.util.itemtouchhelper

import android.graphics.Canvas
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.arman.queuetube.util.itemtouchhelper.adapters.ItemTouchHelperAdapter
import com.arman.queuetube.util.itemtouchhelper.viewholders.ItemTouchHelperViewHolder

class VideoItemTouchHelper(callback: Callback) : ItemTouchHelper(callback) {

    class Callback(private val adapter: ItemTouchHelperAdapter) : ItemTouchHelper.Callback() {

        private var dragFromIndex: Int = 0
        private var dragToIndex: Int = 0

        init {
            this.dragFromIndex = -1
            this.dragToIndex = -1
        }

        override fun isLongPressDragEnabled(): Boolean {
            return true
        }

        override fun isItemViewSwipeEnabled(): Boolean {
            return false
        }

        override fun canDropOver(recyclerView: RecyclerView, current: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            return current.itemViewType == target.itemViewType
        }

        override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
            val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN
            val swipeFlags = ItemTouchHelper.START or ItemTouchHelper.END
            return ItemTouchHelper.Callback.makeMovementFlags(dragFlags, swipeFlags)
        }

        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            if (viewHolder.itemViewType != target.itemViewType) {
                return false
            }

            val fromIndex = viewHolder.adapterPosition
            val toIndex = target.adapterPosition

            if (this.dragFromIndex < 0) {
                this.dragFromIndex = fromIndex
            }
            this.dragToIndex = toIndex

            this.adapter.onItemDragged(viewHolder.adapterPosition, target.adapterPosition, false)
            return true
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
            this.adapter.onItemSwiped(viewHolder.adapterPosition)
        }

        override fun onChildDraw(c: Canvas, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
            if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
                val alpha = 1f - Math.abs(dX) / viewHolder.itemView.width.toFloat()
                viewHolder.itemView.alpha = alpha
                viewHolder.itemView.translationX = dX
            } else {
                super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
            }
        }

        override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
            if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
                if (viewHolder is ItemTouchHelperViewHolder) {
                    (viewHolder as ItemTouchHelperViewHolder).onItemSelected()
                }
            }

            super.onSelectedChanged(viewHolder, actionState)
        }

        override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
            super.clearView(recyclerView, viewHolder)

            if (this.dragFromIndex >= 0 && this.dragFromIndex != this.dragToIndex) {
                this.adapter.onItemDragged(this.dragFromIndex, this.dragToIndex, true)
            }
            this.dragToIndex = -1
            this.dragFromIndex = this.dragToIndex

            viewHolder.itemView.alpha = 1f
            if (viewHolder is ItemTouchHelperViewHolder) viewHolder.onItemClear()
        }

    }

}
